### ISSUE TITLE
Closed world seen acc of MIT-States in the paper is lower than what I have reimplemented

### ISSUE THREAD

**Issue by wqshmzh** [ExplainableML/czsl#20]
Dear author,
     It is a great pleasure of reading your work. Open world compositional zero shot learning is surely an interesting topic worth to be further studied. To make myself more into the algorithm in this paper, I ran the codes several times. However I cannot get the reported results where seen acc was always 10%-20% higher than unseen acc. I found that the codes in evaluation section ran slowly somehow so I reimplemented all the evaluation codes. I used hyper-parameters precisely similar with that in the paper and freezed feature extractor ResNet18 pre-trained in ImageNet. When I ran the algorithm CompCos in closed world setting using MIT-STATES, I got 42.78% seen acc, 28.53% unseen acc in val set and 54.63% seen acc, 25.37% unseen acc in test set. 25.37% unseen acc in test set is the closest to the 24.6% unseen acc reported in the paper. Over-fitting happens both in val and test set. I wonder why seen acc in the paper of 25.3% is very close to unseen acc of 24.6% with only 0.7% difference. What is the test setting and how can I tune the results? 
Thank you!
     Great love for you

**Comment by mancinimassimiliano** [comment]
Hi @wqshmzh,

These results are indeed surprising. However, note that our results are in line with what was reported by previous works (e.g. Table 2 [here](https://arxiv.org/pdf/2004.00587.pdf)), thus I think there might be some differences in how the accuracies are computed. If you can share with me your evaluation code, I would be happy to double-check.

**Comment by wqshmzh** [comment]
Very happy to get your reply! ^_^
1. Some file names in metadata_compositional-split-natural.t7 cannot match the file names in the dataset I downloaded due to some symbols like '\', '/', '*', '?'. So that  I reloaded all 34562 training images, 18375 val images, and 23306 test images according to the train_pairs.txt, val_pairs.txt, and test_pairs.txt.
2. All model related codes are not edited.
3. In evalution for val or test test, for seen acc, 
    S1: I set calibration factor from -1 to 1 with 0.1 stepsize for all unseen pairs in search space (scores after softmax of all unseen labels += calibration factor)   
    S2: For each calibration factor, I picked out images whose pairs (labels) belong to train_pairs.txt as seen images
    S3: Compute classification acc for all seen images
    S4: After cycling through all calibration factors from -1 to 1, select the maximum classification acc of seen images as seen acc.
For unseen acc, 
    For each calibration factor, I picked out rest of all images (images except seen images) as unseen images and compute their classification acc. After cycling through all calibration factors, the maximum classification acc of unseen images is unseen acc.

**Comment by wqshmzh** [comment]
def __call__(self, scores, topk=1):
        '''
        Inputs
            scores: Output scores (num of val_images or test_images, all_pairs) = (18371 or 23306 x1962)
            topk: number of highest class scores for each image
        '''
        num_imgs = scores.shape[0]
        # ======================= AUC ========================= #
        seen_acc_hist, unseen_acc_hist, attr_acc_hist, obj_acc_hist = [], [], [], []
        calibration_factors = torch.arange(-1, 1, 0.1).to(self.device)
        for calibration_factor in calibration_factors:
            new_scores = scores.clone()
            new_scores[:, self.unseen_pairs_idx] += calibration_factor
            # ================== Seen Accuracy ===================== #
            seen_scores = new_scores[self.seen_images_idx]
            # seen_scores = seen_scores[:, self.seen_pairs_idx]
            _, pair_pred = seen_scores.topk(topk, dim=1)
            seen_acc = self.compute_acc(pair_pred, self.seen_images_gt, topk, len(self.seen_images_gt))
            seen_acc_hist.append(seen_acc)
            # ================= Unseen Accuracy ==================== #
            unseen_scores = new_scores[self.unseen_images_idx]
            # unseen_scores = unseen_scores[:, self.unseen_pairs_idx]
            _, pair_pred = unseen_scores.topk(topk, dim=1)
            unseen_acc = self.compute_acc(pair_pred, self.unseen_images_gt, topk, len(self.unseen_images_gt))
            unseen_acc_hist.append(unseen_acc)
            del new_scores
        seen_acc_hist, unseen_acc_hist = torch.tensor(seen_acc_hist).to(self.device), torch.tensor(unseen_acc_hist).to(self.device)
        AUC = torch.trapz(seen_acc_hist, unseen_acc_hist).item()
        # ================== Harmonic Mean ==================== #
        numbers = torch.stack([seen_acc_hist, unseen_acc_hist], dim=1).to(self.device)
        HM_hist = 1. / (1. / numbers).mean(dim=1)
        HM, bestHMidx = HM_hist.unsqueeze(1).max(0)
        HM, bestHMidx = HM.item(), int(bestHMidx)
        # ================== Best Seen Accuracy ==================== #
        seen_acc = seen_acc_hist.max().item()
        # ================= Best Unseen Accuracy =================== #
        unseen_acc = unseen_acc_hist.max().item()
        
        # return attr_acc, obj_acc, seen_acc, unseen_acc, HM, AUC
        return seen_acc, unseen_acc, HM, AUC

def compute_acc(self, pair_pred, all_gt, topk, num_imgs):
        pair_pred = pair_pred.contiguous().view(-1)
        all_gt = torch.stack([all_gt]*topk, dim=1).contiguous().view(-1).to(self.device)
        acc = (pair_pred == all_gt).float().sum().item() / num_imgs
        return acc

**Comment by mancinimassimiliano** [comment]
Thanks, @wqshmzh!
I skimmed through the code and I did not notice anything in particular. I will have a deeper look in the next few days and let you know. Before that, the only two things that I noticed from your description are:

> Some file names in metadata_compositional-split-natural.t7 cannot match the file names in the dataset I downloaded due to some symbols like '', '/', '*', '?'. So that I reloaded all 34562 training images, 18375 val images, and 23306 test images according to the train_pairs.txt, val_pairs.txt, and test_pairs.txt.

Maybe a question here: are the number of training, validation, and test images the same? I ask this because `metadata_compositional-split-natural.t7` contains the image splits, so data should be divided accordingly to that.

**Comment by mancinimassimiliano** [comment]
I think I was able to obtain your data numbers. Practically, if I loop over `metadata_compositional-split-natural.t7` and filter by composition (i.e. `test_pairs.txt`) rather than the set (i.e. train vs validation vs test), I got 23'306 test images. 

However, note that `test_pairs.txt` contains also the `seen pairs` contained in the training set, thus there are 10'311 images that are seen both at training and at test time. this should explain the gap in performance.

**Comment by wqshmzh** [comment]
Sorry I'm late to come due to some busy work. I apologize! Thanks for your reply! Difference of data division is truely the key to explain the gap. Phew! My evalution code is correct and I can use it in my following czsl algorithms. Happy with that. So would you mind sharing the way on how the file metadata_compositional-split-natural.t7 is constructed (i.e. How can one output the file metadata_compositional-split-natural.t7 on his own?). Oh BTW, I see that the calibration factor you mentioned is model dependent. What methed did you use to select model dependent calibration factor? It would be nice if you could share the related paper with us! Hahaha

**Comment by mancinimassimiliano** [comment]
Hello @wqshmzh!

The answer is that it is not possible: `metadata_compositional-split-natural.t7` says which image belongs to which split. The information on the splits is thus contained only on `metadata_compositional-split-natural.t7` and this file is provided by the authors of the datasets/benchmarks. 

The files `train_pairs.txt`, `val_pairs.txt`, and `test_pairs.txt` say which pairs are present in each split, but **without** any information on the images. 

Thus, by looking at the ground-truth pairs per split in `metadata_compositional-split-natural.t7`, one can extract the files `train_pairs.txt`, `val_pairs.txt`, and `test_pairs.txt`. However, the reverse is not possible since those files contain information on which pairs are present in each split but not which subset of images of those pairs (this is why the train/val/test overlap would arise for the overlapping seen compositions).
